src/interactive-window/generatedCodeStoreManager.ts

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { inject, injectable } from 'inversify';
+import { IKernel, IKernelProvider } from '../kernels/types';
+import { InteractiveWindowView } from '../notebooks/constants';
+import { getAssociatedNotebookDocument } from '../notebooks/controllers/kernelSelector';
+import { IExtensionSyncActivationService } from '../platform/activation/types';
+import { disposeAllDisposables } from '../platform/common/helpers';
+import { IDisposable, IDisposableRegistry } from '../platform/common/types';
+import { CellHashProviderFactory } from './editor-integration/cellHashProviderFactory';
+import { IGeneratedCodeStorageFactory } from './editor-integration/types';
+
+@injectable()
+export class GeneratedCodeStorageManager implements IExtensionSyncActivationService {
+    private readonly disposables: IDisposable[] = [];
+    constructor(
+        @inject(IKernelProvider) private readonly kernelProvider: IKernelProvider,
+        @inject(IDisposableRegistry) disposables: IDisposableRegistry,
+        @inject(CellHashProviderFactory) private readonly hashProvider: CellHashProviderFactory,
+        @inject(IGeneratedCodeStorageFactory) private readonly storageFactory: IGeneratedCodeStorageFactory
+    ) {
+        disposables.push(this);
+    }
+    dispose() {
+        disposeAllDisposables(this.disposables);
+    }
+    activate(): void {
+        this.kernelProvider.onDidCreateKernel(this.onDidCreateKernel, this, this.disposables);
+    }
+    private onDidCreateKernel(kernel: IKernel) {
+        const nb = getAssociatedNotebookDocument(kernel);
+        if (kernel.creator !== 'jupyterExtension' || nb?.notebookType !== InteractiveWindowView) {
+            return;
+        }
+        // Possible we changed kernels for the same document.
+        this.storageFactory.get(nb).clear();
+
+        kernel.onRestarted(
```

Comment:
This class will be responsible for clearing the hashes & the like.
Still need to wire up generation of the hashes.